SIPS/sip-20.md

```diff
@@ -0,0 +1,181 @@
+---
+sip: 20
+title: External data entry point
+status: Draft
+author: David Drazic (@david0xd)
+created: 2023-12-15
+---
+
+## Abstract
+
+This SIP proposes a new permission that enables Snaps to receive data from external sources.
+
+In the initial version, this SIP proposes support for WebSockets, but it can be extended to support other data sources like blockchain events in the future.
+The Snap can specify the external data source in the Snap manifest. The client then connects to the external data source and sends the data to the Snap.
+The Snap can then do whatever it wants with the data. This initial version only supports one-way communication from the external source to the Snap. The Snap can't send any data back to the external source.
+
+## Motivation
+
+Snaps are currently limited in their ability to receive data from external sources; they have to rely on user actions or cron jobs to fetch data, so they can't react to events in real time. Snaps also cannot use WebSocket connections to receive data from external sources, and are limited to HTTP requests.
+
+## Specification
+
+> Formal specifications are written in Typescript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+### Snap manifest
+
+This SIP introduces a new permission named `endowment:external-data`.
+
+```typescript
+/**
+ * A WebSocket connection as specified in the Snap manifest.
+ *
+ * @property type - The type of connection. Currently, only WebSockets are
+ * supported.
+ * @property url - The URL to connect to. This must be a valid WebSocket URL,
+ * starting with `wss://`. URLs starting with `ws://` are not allowed.
+ * @property dataType - The type of data expected from the WebSocket. If this
+ * field is omitted, text is assumed.
+ */
+type WebSocketConnection = {
+  type: 'websocket';
+  url: string;
+  dataType?: 'text' | 'binary';
+};
+
+/**
+ * An external data connection as specified in the Snap manifest.
+ *
+ * Currently, only {@link WebSocketConnection} is supported.
+ */
+type ExternalDataConnection = WebSocketConnection;
+
+/**
+ * External data connections as specified in the Snap manifest.
+ *
+ * This is the value of the `endowment:external-data` field in the Snap
+ * manifest.
+ */
+type ExternalData = ExternalDataConnection[];
+```
+
+#### Example
+
+The new field can be specified as follows in a `snap.manifest.json` file:
+
+```json
```

Comment:
We may want to describe this as an object to more easily add more keys to the caveat.

For example, if we want to support SIP-21 here.

```json
{
  "initialPermissions": {
    "endowment:external-data": {
      "maxRequestTime": 50000,
      "connections": [
        {
          "type": "websocket",
          "url": "wss://example.com/binary/endpoint",
          "dataType": "binary"
        },
        {
          "type": "websocket",
          "url": "wss://example.com/text/endpoint",
          "dataType": "text"
        }
      ]
    }
  }
}
```

Reply:
Updated as you suggested.

Reply:
Should probably be added to the definition then?

Reply:
Added as you suggested and answered here: https://github.com/MetaMask/SIPs/pull/125#discussion_r1494704575